base/compiler/typelimits.jl

```diff
@@ -515,8 +515,12 @@ function tmerge(lattice::PartialsLattice, @nospecialize(typea), @nospecialize(ty
                     tyi = ai
                 elseif is_lattice_equal(lattice, bi, ft)
                     tyi = bi
+                elseif (tyi′ = tmerge_field(lattice, ai, bi); tyi′ !== nothing)
+                    # allow external lattice implementation to provide a custom field-merge strategy
+                    tyi = tyi′
```

Comment:
```suggestion
                    if issimplertype(tyi′, ai) && issimplertype(tyi′, bi)
                        tyi = tyi′
                    else
                        tyi = Any # the custom tmerge_field implementation violates the requirements of this lattice: abort
                    end
```

Reply:
That doesn't seem right. The requirement is that the total tmerge you can get is bounded, not that any individual tmerge produces something simpler than the things you're merging.

Reply:
I agree that an external `tmerge_field` should apply some complexity-limiting mechanisms, such as limiting internal Union-type to split up to 4, but currently we don't have a good way to detect if such a limitation has been applied to individual operations.

Reply:
In practice, we have found that to be a poor heuristic for Unions, since it is too big for practical purposes of similar things and too small for distinct things. It also does not converge in total, unless you had also applied some complexity bounding to each individual part of the Union